pytorch_lightning/loops/training_batch_loop.py

```diff
@@ -0,0 +1,722 @@
+
+    def _num_training_batches_reached(self, is_last_batch: bool = False) -> bool:
+        """Checks whether sufficient training batches have been processed.
+
+        Args:
+            is_last_batch: Whether the current batch is the last one
+        """
+        # TODO(@awaelchli): use progress tracking of batches instead of manual batch_idx
+        return (self.batch_idx + 1) == self.trainer.num_training_batches or is_last_batch
+
+    def should_accumulate(self) -> bool:
+        """Checks if the optimizer step should be performed or gradients should be accumulated for the current step."""
+        # checks if backward or backward + optimizer step (via closure)
+        accumulation_done = self._accumulated_batches_reached()
+        is_final_batch = self._num_training_batches_reached()
+        return not (accumulation_done or is_final_batch)
+
+    def tbptt_split_batch(self, batch: Any) -> List[Any]:
+        """Splits a single batch into a list of sequence steps for tbptt.
+
+        Args:
+            batch: the current batch to split
+
+        """
+        splits = [batch]
+        if self.trainer.truncated_bptt_steps is not None:
+            model_ref = self.trainer.lightning_module
+            with self.trainer.profiler.profile("tbptt_split_batch"):
+                splits = model_ref.tbptt_split_batch(batch, self.trainer.truncated_bptt_steps)
+        return splits
+
+    def build_train_args(self, batch: Any, batch_idx: int, opt_idx: int, hiddens: Tensor) -> List[Any]:
+        """Builds arguments for train step
+
+        Args:
+            batch: the current batch to train on
+            batch_idx: the index of the current batch
+            opt_idx: the index of the current optimizer
+            hiddens: the hidden state of the previous RNN iteration
+
+        Returns:
+            the positional arguments for training
+        """
+        # enable not needing to add opt_idx to training_step
+        args = [batch, batch_idx]
+
+        if len(self.trainer.optimizers) > 1:
+            if self.trainer.has_arg("training_step", "optimizer_idx"):
+                if not self.trainer.lightning_module.automatic_optimization:
+                    self.warning_cache.warn(
+                        "`training_step` hook signature has changed in v1.3."
+                        " `optimizer_idx` argument has been removed in case of manual optimization. Support for"
+                        " the old signature will be removed in v1.5", DeprecationWarning
+                    )
+                args.append(opt_idx)
+            elif not self.trainer.has_arg(
+                "training_step", "optimizer_idx"
+            ) and self.trainer.lightning_module.automatic_optimization:
+                raise ValueError(
+                    f"Your LightningModule defines {len(self.trainer.optimizers)} optimizers but"
+                    ' `training_step` is missing the `optimizer_idx` argument.'
+                )
```

Comment:
```suggestion
        model_auto_optim = self.trainer.lightning_module.automatic_optimization
        if len(self.trainer.optimizers) > 1:
            if self.trainer.has_arg("training_step", "optimizer_idx"):
                if not model_auto_optim:
                    self.warning_cache.warn(
                        "`training_step` hook signature has changed in v1.3."
                        " `optimizer_idx` argument has been removed in case of manual optimization. Support for"
                        " the old signature will be removed in v1.5", DeprecationWarning
                    )
                args.append(opt_idx)
            elif not self.trainer.has_arg(
                "training_step", "optimizer_idx"
            ) and model_auto_optim:
                raise ValueError(
                    f"Your `LightningModule` defines {len(self.trainer.optimizers)} optimizers but"
                    ' `training_step` is missing the `optimizer_idx` argument.'
                )
```

Reply:
hard to see what's changed, did you define model_auto_optim?
This code we haven't touched, it comes from old loop
I prefer not to mess with the if-else logic.